benchmarks/attention/benchmark_attention.py

```diff
@@ -0,0 +1,193 @@
+# Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+import os, sys, time
+import pandas as pd
+import numpy as np
+import torch
+import nvtx
+import transformer_engine
+from tests.pytorch.fused_attn.test_fused_attn import (
+    ModelConfig,
+    _is_flash_attention_supported,
+    _is_fused_attention_supported,
+    _is_unfused_attention_supported,
+    _run_dot_product_attention
+)
+
+pd.set_option("display.precision", 4)
+
+# data type
+dtype = torch.bfloat16
+# number of iterations after 3 warmup iterations
+num_iters = 3
+# checkpointing
+ckpt_attn = False
+# workspace optimization path for cuDNN attention
+workspace_opt = True
+# QKV memory layout
+qkv_layout = 'bshd_bshd_bshd'
+# sliding window attention
+swa = False
+# padding between sequences for qkv_format=thd
+pad_between_seqs = False
+# training mode
+is_training = True
+
+model_configs = {
+    #   test:             b,  h, hg,   d,   sq,  skv,   p,     mask,              bias
+    "test_0": ModelConfig(2, 16, 16,  64,  512,  512, 0.0, "no_mask",         "no_bias"), # short seq
+    "test_1": ModelConfig(2, 16, 16, 128, 2048, 2048, 0.0,  "causal",         "no_bias"), # longer seq, mask
+    "test_2": ModelConfig(2, 16, 16, 128, 2048, 2048, 0.0,  "causal", "post_scale_bias"), # bias
+    "test_3": ModelConfig(2, 32,  4, 128, 8192, 8192, 0.0,  "causal",         "no_bias"), # GQA
+}
+
+def benchmark_dot_product_attention(model, fused_attn_supported, flash_attn_supported):
+    config = model_configs[model]
+    if dtype == torch.bfloat16:
+        tols = dict(atol=2.5e-2, rtol=2.5e-2) 
+    else:
+        tols = dict(atol=5e-3, rtol=5e-3)
+
+    cudnn_times = []
+    flash_times = []
+    warmup_iters = 3
+    for i in range(num_iters+warmup_iters):
+        if i == warmup_iters:
+            torch.cuda.cudart().cudaProfilerStart()
+
+        if fused_attn_supported:
+            torch.cuda.synchronize()
+            fused_attn_start = time.time()
+            fused_attn_fwd, fused_attn_bwd = _run_dot_product_attention(
```

Comment:
We usually like to run several times then add the synchronize at the end and computes the average when benchmarking. This can overlap the CPU launch time and GPU compute to reflect the real workload. Adding synchronize per iter will measure the CPU launch/pre-processing time as those time is usually overlapped in the real workload.